### PR TITLE
Add forcename option to API

### PR DESCRIPTION
**Pull request by AdrienClairembault**
Some API requests returns ids (e.g. users_id) that a self-service user can't work with (as he is not allowed to query /User/{id}) and thus can't get the name matching this id (e.g. the name of the author of an ITILFollowup).

This proposal add a "forcename" option that allow to access the rawname of the items.  

Example request: 
```http
GET apirest.php/ITILFollowup?id=22
&forcename[0]=users_id
&forcename[1]=users_id_editor
```

Response:
```json
{
        "id": 22,
        "itemtype": "Ticket",
        "items_id": 5,
        "date": "2019-02-25 16:27:05",
        "users_id": 2,
        "users_id_editor": 2,
        "content": "&lt;p&gt;aaaqdzdqdqzdqzdqzdqz&lt;/p&gt;&lt;p&gt;&lt;a href=\"http://72.glpi-fusion-94.local/front/document.send.php?docid=4&amp;tickets_id=5\" target=\"_blank\" rel=\"noopener\"&gt;&lt;img src=\"http://72.glpi-fusion-94.local/front/document.send.php?docid=4&amp;tickets_id=5\" alt=\"3577bb62-c6a01f19-5c740943affc40.83487561\" width=\"185\" /&gt;&lt;/a&gt;&lt;/p&gt;",
        "is_private": 0,
        "requesttypes_id": 1,
        "date_mod": "2020-01-28 10:06:09",
        "date_creation": "2019-02-25 16:27:05",
        "timeline_position": 1,
        "sourceitems_id": 0,
        "sourceof_items_id": 0,
 |      "_names": {
 |          "users_id": "glpi",
 |          "users_id_editor": "glpi",
 |          "links": [
 |              {
 |                  "rel": "User",
 |                  "href": "http://glpi-fusion-94.local/api/User/glpi"
 |              },
 |              {
 |                  "rel": "User",
 |                  "href": "http://glpi-fusion-94.local/api/User/glpi"
 |              }
 |          ]
 |      }
}
```

Performance impact:

Benchmark request (80 followups, two names requested):
```http
GET apirest.php/ITILFollowup?range=0-80
&forcename[0]=users_id
&forcename[1]=users_id_editor
```

Without cache:
```
PHP message: Additional time: 76.05 ms
PHP message: Average time per name fetched:  0.93 ms
```

With cache:
```
PHP message: Additional time: 30.44 ms
PHP message: Average time per name fetched:  0.37 ms
```

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

**Comment by AdrienClairembault** [comment]
Tests / docs will be added after the first wave of reviews if they are positive.

**Comment by AdrienClairembault** [comment]
@cedric-anne 

As we discussed yesterday I agree with you on this except on one point:
> makes CommonDBTM::getFriendlyName() use cache when it is available or call CommonDBTM::computeFriendlyName() to init its cache when needed.

CommonDBTM::getFriendlyName() should set the cache but not use it.
Building the friendly name has no performance cost, what slow down the API is to load the object from DB before being able to get his friendly name.

This mean that the API should still check the cache to avoid loading the object if his friendly name is cached (and CommonDBTM::getFriendlyName should set the friendly name cache so the API can benefit from already cached names even if it was not the one who used them).

File: inc/api.class.php [hunk 1]
```diff
@@ -1308,6 +1337,42 @@ protected function getItems($itemtype, $params = [], &$totalcount = 0) {
...
+      $_names = [];
+
+      foreach ($params[self::PARAM_FORCENAME] as $fn_fkey) {
+         if ($fn_fkey == "self") {
+            // Get friendlyname for current item
+            $fn_itemtype = $self_itemtype;
+            $fn_id = $data["id"];
+         } else {
+            // Get friendlyname for given fkey
+            $fn_itemtype = getItemtypeForForeignKeyField($fn_fkey);
```
Comment (cedric-anne) [inline inc/api.class.php]: You should validate that this is a valid itemtype and classname, for example by using `getItemForItemtype`.

File: inc/commondbtm.class.php [hunk 2]
```diff
@@ -5263,4 +5267,77 @@ private function assetBusinessRules($condition) {
...
+   /**
+    * Return the computed friendly name and set the cache.
+    *
+    * @since 9.5
+    *
+    * @return string
+    */
+   final public function getFriendlyName() {
+      global $GLPI_CACHE;
+
```
Comment (cedric-anne) [inline inc/commondbtm.class.php]: I would also check for cache hit here.

File: inc/commondbtm.class.php [hunk 3]
```diff
@@ -3414,16 +3417,17 @@ static function getCompleteNameField() {
    }
 
 
-   /** Get raw name of the object
-    * Maybe overloaded
+   /**
+    * Deprecated, replaced by getFriendlyName
     *
+    * @deprecated
```
Comment (cedric-anne) [inline inc/commondbtm.class.php]:
```suggestion
    * @deprecated 9.4.6
```

File: inc/commondbtm.class.php [hunk 4]
```diff
@@ -3414,16 +3417,17 @@ static function getCompleteNameField() {
...
     * @see CommonDBTM::getNameField
-    *
     * @since 0.85
     *
     * @return string
    **/
    function getRawName() {
+      \Toolbox::deprecated();
```
Comment (cedric-anne) [inline inc/commondbtm.class.php]:
```suggestion
      \Toolbox::deprecated('Use CommonDBTM::getFriendlyName()');
```

File: inc/commondbtm.class.php [hunk 5]
```diff
@@ -3414,16 +3417,17 @@ static function getCompleteNameField() {
    }
 
 
-   /** Get raw name of the object
-    * Maybe overloaded
+   /**
+    * Deprecated, replaced by getFriendlyName
```
Comment (cedric-anne) [inline inc/commondbtm.class.php]: No need to rewrite phpdoc.

File: inc/commondbtm.class.php [hunk 6]
```diff
@@ -3414,16 +3417,17 @@ static function getCompleteNameField() {
...
     * @see CommonDBTM::getNameField
-    *
     * @since 0.85
     *
     * @return string
    **/
    function getRawName() {
+      \Toolbox::deprecated();
 
       if (isset($this->fields[static::getNameField()])) {
```
Comment (cedric-anne) [inline inc/commondbtm.class.php]: You should call `$this->getFriendlyName()` here, so removing overrides of `getRawName` will be transparent for end user.

File: inc/commondbtm.class.php [hunk 7]
```diff
@@ -5263,4 +5267,77 @@ private function assetBusinessRules($condition) {
...
+   public static function getCacheKeyForFriendlyName($itemtype, $id) {
+      return "raw_name__{$itemtype}__{$id}";
+   }
+
+   /**
+    * Get friendly name by items id
+    * The purpose of this function is to try to access the friendly name
+    * without having to read the object from the database
+    *
+    * @since 9.5
```
Comment (cedric-anne) [inline inc/commondbtm.class.php]: 9.5 or 9.4.6 ?

File: inc/commonitiltask.class.php [hunk 8]
```diff
@@ -549,11 +549,11 @@ function cleanDBonPurge() {
    // SPECIFIC FUNCTIONS
 
    /**
```
Comment (cedric-anne) [inline inc/commonitiltask.class.php]: You should remove the entire PHPDoc block, as it provides no usefull information.

Same remark for other classes except `CommonDBTM`.

File: inc/operatingsystemkernelversion.class.php [hunk 9]
```diff
@@ -62,10 +62,10 @@ function displaySpecificTypeField($ID, $field = []) {
       }
    }
 
-   function getRawName() {
+   function computeFriendlyName() {
       $kernel = new OperatingSystemKernel();
-      $kname = $kernel->getRawName();
-      $kvname = parent::getRawName();
+      $kname = $kernel->getFriendlyName();
```
Comment (cedric-anne) [inline inc/operatingsystemkernelversion.class.php]: Should not be called on an item not loaded.

File: inc/operatingsystemkernelversion.class.php [hunk 10]
```diff
@@ -62,10 +62,10 @@ function displaySpecificTypeField($ID, $field = []) {
...
+      $kvname = parent::getFriendlyName();
```
Comment (cedric-anne) [inline inc/operatingsystemkernelversion.class.php]: Will loop unless you call something like:
```suggestion
      $kvname = parent::computeFriendlyName();
```

File: inc/commondbtm.class.php [hunk 11]
```diff
@@ -1336,7 +1336,7 @@ function post_addItem() {
     * @return boolean true on success
    **/
    function update(array $input, $history = 1, $options = []) {
-      global $DB;
+      global $DB, $GLPI_CACHE;
```
Comment (cedric-anne) [inline inc/commondbtm.class.php]:
```suggestion
      global $DB;
```

File: inc/commondbtm.class.php [hunk 12]
```diff
@@ -1475,6 +1475,9 @@ function update(array $input, $history = 1, $options = []) {
                QueuedNotification::forceSendFor($this->getType(), $this->fields['id']);
             }
 
+            // Update raw names cache (for API)
+            $this->getFriendlyName();
```
Comment (cedric-anne) [inline inc/commondbtm.class.php]: This will not invalidate cache.

**Comment by AdrienClairembault** [comment]
> Seems OK for me, but should it target 9.4 or 9.5 branch ?

Rebased to 9.5.

**Comment by cedric-anne** [comment]
An entry in CHANGELOG.md should be added for deprecation of `getRawName()`.

**Comment by AdrienClairembault** [comment]
As we discussed @orthagh @cedric-anne :
* forcename => add_keys_names
* _names => _keys_names
* self =-> id

**Comment by rafaeldomi** [comment]
Trying to install a freshly cloned 9.5/bugfixes i get this error:

GLPI SETUP

Step 3
Initialization of the database
Uncaught Exception Error: Call to a member function delete() on null in [..]/glpi95/inc/commondbtm.class.php at line 1455, which points to this line, commited in this pull

```
            // Update raw names cache (for API)
            $GLPI_CACHE->delete(self::getCacheKeyForFriendlyName(
               $this->getType(),
               $this->fields['id']
            ));
            $this->getFriendlyName();
```

**Pull request merged**